### ISSUE TITLE
The dead thread reported in ICorProfilerInfo4::EnumThreads

### ISSUE THREAD

**Issue by ww898** [dotnet/runtime#60566]
### Description

As described in documentation `ICorProfilerInfo::ForceGC` requires separate native thread for working. The thread will be finished by my profiler during detaching procedure. And I see it in lldb. However, after attaching again to the process the `ICorProfilerInfo4::EnumThreads` call returns me the already dead thread where `ICorProfilerInfo::ForceGC` was called sometime ago .

It's the important problem because the thread id can be already reused for a new native thread.

My investigation shows the CoreCLR keeps the static list of managed threads. and only alive threads will be returned in `ICorProfilerInfo4::EnumThreads`. On the other hand the `ICorProfilerInfo::ForceGC` adds current thread in this list.

### Reproduction Steps

- Create following "hello world" application:
```
using System;
namespace test { class Program { static void Main(string[] args) {
  Console.WriteLine("Hello World!");
  Console.ReadLine();
}}}
```
- Publish it as a self contained application and run it in separate console
- Attach and detach a profiler which use `ICorProfilerInfo::ForceGC` and dump somewhere the thread id after the call.
- Attach lldb and dump the thread list. Check that the thread id is not in the list.
- Attach a profiler which use `ICorProfilerInfo4::EnumThreads` and see that the dead thread id is in the list of managed threads.

### Actual behavior

The native thread id which interesting us is `1BCF69`.

The part of the first profiler log;
```
00000000001BCF69 0 0000000000000000 GCCQ::ForceGcThread
00000000001BCF69 0 0000000000000000 GCCQ::ForceGc // <-- ICorProfilerInfo::ForceGC started
00000000001BCF69 1 000000014180AA00 MProf::GarbageCollectionStarted gcNumber=0 reason=induced c=5
00000000001BCF69 1 000000014180AA00 GCCQ::OnStarted gcNumber=0 gcState=normal
00000000001BCF69 1 000000014180AA00 MDG2::FlushMetadata
00000000001BCF69 1 000000014180AA00 MDG2::FlushMetadata res=ok
00000000001BCF69 1 000000014180AA00 MProf::GarbageCollectionStarted gcNumber=0 res=ok
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122F650
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122FA50
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122FC40
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122FCC0
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122FD40
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122FDC0
00000000001BCF69 1 000000014180AA00 MProf::FinalizeableObjectQueued f=C oid=000000010122FE48
00000000001BCF69 1 000000014180AA00 MProf::SurvivingReferences2 c=1
00000000001BCF69 1 000000014180AA00 MProf::SurvivingReferences2 c=91
00000000001BCF69 1 000000014180AA00 MProf::RootReferences2 c=42
00000000001BCF69 1 000000014180AA00 BProf::RootReferences c=42
00000000001BCF69 1 000000014180AA00 MProf::ObjectReferences oid=0000000101220048 cid=0000000280E48C80 c=0
00000000001BCF69 1 000000014180AA00 MProf::GarbageCollectionFinished gcNumber=0
00000000001BCF69 1 000000014180AA00 MProf::GarbageCollectionFinished gcNumber=0 res=ok
00000000001BCF69 1 000000014180AA00 GCCQ::OnFinished gcNumber=0 gcState=normal
00000000001BCF69 0 0000000000000000 GCCQ::ForceGc res=ok  // <-- ICorProfilerInfo::ForceGC finished
00000000001BCF69 0 0000000000000000 GCCQ::ForceGcThread res=ok
```

lldb check:
```
(lldb) thread list
Process 52251 stopped
* thread #1: tid = 0x1bce01, 0x0000000194e1a538 libsystem_kernel.dylib`read + 8, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  thread #2: tid = 0x1bce03, 0x0000000194e19954 libsystem_kernel.dylib`mach_msg_trap + 8
  thread #3: tid = 0x1bce04, 0x0000000194e1f8cc libsystem_kernel.dylib`kevent + 8
  thread #4: tid = 0x1bce05, 0x0000000194e21ad0 libsystem_kernel.dylib`poll + 8
  thread #5: tid = 0x1bce06, 0x0000000194e1a3d4 libsystem_kernel.dylib`__open + 8
  thread #6: tid = 0x1bce07, 0x0000000194e1d0c0 libsystem_kernel.dylib`__psynch_cvwait + 8
  thread #7: tid = 0x1bce08, 0x0000000194e1d0c0 libsystem_kernel.dylib`__psynch_cvwait + 8
  thread #8: tid = 0x1bce0a, 0x0000000194e1a538 libsystem_kernel.dylib`read + 8
  thread #9: tid = 0x1bcf5c, 0x0000000194e1d0c0 libsystem_kernel.dylib`__psynch_cvwait + 8
  thread #10: tid = 0x1bcf5d, 0x0000000194e1b604 libsystem_kernel.dylib`__workq_kernreturn + 8
```

The part of the second profiler log:
```
00000000001BCE05 1 0000000000000000 BProf::ProfilerAttachComplete // <--- ICorProfilerCallback3::ProfilerAttachComplete
00000000001BCE05 1 0000000000000000 BProf::ProfilerAttachComplete tid=000000015200C000
00000000001BCE05 1 0000000000000000 BPPool::DoThreadRecovered tid=000000015200C000 osTid=001BCE01
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=000000015200C000 osTid=001BCE01 (1822209)
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=000000015200C000 id=00000001
00000000001BCE05 1 0000000000000000 BProf::ProfilerAttachComplete tid=000000015200C600
00000000001BCE05 1 0000000000000000 BPPool::DoThreadRecovered tid=000000015200C600 osTid=001BCE08
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=000000015200C600 osTid=001BCE08 (1822216)
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=000000015200C600 id=00000002
00000000001BCE05 1 0000000000000000 BProf::ProfilerAttachComplete tid=0000000131808200
00000000001BCE05 1 0000000000000000 BPPool::DoThreadRecovered tid=0000000131808200 osTid=001BCE0A
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=0000000131808200 osTid=001BCE0A (1822218)
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=0000000131808200 id=00000003
00000000001BCE05 1 0000000000000000 BProf::ProfilerAttachComplete tid=000000014180AA00 // <--- The dead thread reported here!!!
00000000001BCE05 1 0000000000000000 BPPool::DoThreadRecovered tid=000000014180AA00 osTid=001BCF69
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=000000014180AA00 osTid=001BCF69 (1822569)
00000000001BCE05 1 0000000000000000 BPool::DoThreadRecovered tid=000000014180AA00 id=00000004
00000000001BCE05 1 0000000000000000 ERROR: No such thread
```

### Configuration

I reproduced the issue on macOS:
- Catalina x64: .NET v5.0.402 x64
- Big Sur x64: .NET v5.0.402 x64 / v6.0-rc2 x64
- Monterey arm64: .NET v5.0.402 x64 + Rosetta2 / v6.0-rc2 x64 + Rosetta2 / v6.0-rc2 arm64

The Linux x64/arm64 versions work properly.

**Comment by tommcdon** [comment]
@davmason

**Comment by ww898** [comment]
I have just reproduced the issue on macOS Catalina x64.

**Comment by davmason** [comment]
Hi @ww898, thanks for reporting this. Would you be able to create a simplified repro of a profiler that just creates the GC thread and then iterates the threads?

**Comment by janvorli** [comment]
@davmason a profiler repro would be nice

**Comment by janvorli** [comment]
What I was trying to say is that I don't see anything in our code that would prevent us from returning an already dead thread.

**Comment by ww898** [comment]
.NET 6.0 rc2 macOS Monterey arm64

I see `Saw ForceGC thread id in enum list!` in logs:
```
% ./repro.sh
Warning NETSDK1174: The abbreviation of -p for --project is deprecated. Please use --project.
Building...
Initialize profiler!
Launching ForceGC thread
Thread created, id=48838a00
Thread created, id=4902b800
Thread created, id=4902da00
Trying to force gc
Thread created, id=38808200
GC forced
ForceGC threadId=38808200
Thread created, id=3900a200
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread created, id=3900ae00
Thread created, id=4889b800
Thread created, id=39014a00
Thread created, id=48892a00
Thread created, id=48897600
Thread created, id=3901e200
Thread created, id=4904b000
Thread created, id=4904be00
Thread created, id=488a5e00
Thread created, id=4904cc00
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
warn: Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer[5]
      The application is trying to access the ASP.NET Core developer certificate key. A prompt might appear to ask for permission to access the key. When that happens, select 'Always Allow' to grant 'dotnet' access to the certificate key in the future.
fail: Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer[4]
      The ASP.NET Core developer certificate is in an invalid state. To fix this issue, run the following commands 'dotnet dev-certs https --clean' and 'dotnet dev-certs https' to remove all existing ASP.NET Core development certificates and create a new untrusted developer certificate. On macOS or Windows, use 'dotnet dev-certs https --trust' to trust the new certificate.
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:5001
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /Users/jetbrains/work/PROF-1182/macos_thread_profiler_repro/webapp
Thread destroyed, id=4902da00
Thread created, id=39808200
Thread destroyed, id=4904cc00
Thread destroyed, id=488a5e00
Thread destroyed, id=4889b800
Thread destroyed, id=48892a00
Thread destroyed, id=4904be00
Thread destroyed, id=4904b000
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread destroyed, id=39808200
Thread created, id=39808200
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread destroyed, id=39808200
^CThread created, id=18808200
^CCalling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread created, id=18809600
Thread created, id=1880aa00
Thread created, id=39808200
Thread created, id=39809a00
Thread created, id=29008c00
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
Thread destroyed, id=39808200
Thread destroyed, id=1880aa00
Thread created, id=39809000
Thread created, id=488a0200
Thread created, id=4904aa00
Thread created, id=3902ec00
Thread created, id=3880ae00
Thread created, id=3880bc00
Thread created, id=49048c00
Thread destroyed, id=3900ae00
```

**Comment by janvorli** [comment]
Right, that's what I see too. The reason for those is what I've mentioned above in https://github.com/dotnet/runtime/issues/60566#issuecomment-950026097 . But I got no SIGSEGVs when running under a debugger. Do you still get those like with your profiler?

**Comment by ww898** [comment]
I have just finished debugging .NET 5.0.11 macOS Big Sur x64 where the situation is the same. The `Thread::DetachThread()` in `EEDllMain()` never called for `ForceGC` thread because the condition in +193..+196:
```
    0x10246a591 <+97>:  callq  0x10200afe0               ; PAL_GetNativeExceptionHolderHead
    0x10246a596 <+102>: movq   %rax, -0x38(%rbp)
    0x10246a59a <+106>: movq   (%rax), %rcx
    0x10246a59d <+109>: movq   %rcx, -0x30(%rbp)
    0x10246a5a1 <+113>: leaq   -0x40(%rbp), %rcx
    0x10246a5a5 <+117>: movq   %rcx, (%rax)
    0x10246a5a8 <+120>: movq   -0x48(%rbp), %rax
    0x10246a5ac <+124>: movl   0x8(%rax), %eax
    0x10246a5af <+127>: testl  %eax, %eax
    0x10246a5b1 <+129>: je     0x10246a618               ; <+232>
    0x10246a5b3 <+131>: cmpl   $0x3, %eax
    0x10246a5b6 <+134>: je     0x10246a5e5               ; <+181>
    0x10246a5b8 <+136>: cmpl   $0x1, %eax
    0x10246a5bb <+139>: jne    0x10246a6bf               ; <+399>
    0x10246a5c1 <+145>: leaq   0xd3fd0(%rip), %rdi       ; g_SystemInfo
    0x10246a5c8 <+152>: callq  0x102023fc0               ; GetSystemInfo
    0x10246a5cd <+157>: leaq   0x4cc(%rip), %rdi         ; GetThreadUICultureNames(StringArrayList*)
    0x10246a5d4 <+164>: leaq   0xb05(%rip), %rsi         ; GetThreadUICultureId(char16_t (*) [85])
    0x10246a5db <+171>: callq  0x10233c400               ; SetResourceCultureCallbacks(int (*)(StringArrayList*), int (*)(char16_t (*) [85]))
    0x10246a5e0 <+176>: jmp    0x10246a6bf               ; <+399>
    0x10246a5e5 <+181>: leaq   0xcbb84(%rip), %rdi       ; gCurrentThreadInfo
    0x10246a5ec <+188>: callq  *(%rdi)
    0x10246a5ee <+190>: movq   (%rax), %rbx
    0x10246a5f1 <+193>: testq  %rbx, %rbx
->  0x10246a5f4 <+196>: je     0x10246a6bf               ; <+399>
```
+399 is the address just after `Thread::DetachThread()`
```
    0x10246a6aa <+378>: movq   %rbx, %rdi
    0x10246a6ad <+381>: callq  0x1022683c0               ; Thread::RareEnablePreemptiveGC()
    0x10246a6b2 <+386>: movq   %rbx, %rdi
    0x10246a6b5 <+389>: movl   $0x1, %esi
    0x10246a6ba <+394>: callq  0x1021493a0               ; Thread::DetachThread(int)
    0x10246a6bf <+399>: leaq   0xaa84a(%rip), %rax       ; vtable for NativeExceptionHolderBase + 16
    0x10246a6c6 <+406>: movq   %rax, -0x40(%rbp)
    0x10246a6ca <+410>: movq   -0x38(%rbp), %rax
```

`GetThread()` implementation
```
inline Thread* GetThread()
{
    Thread* pThread = gCurrentThreadInfo.m_pThread;
    _ASSERTE(pThread);
    return pThread;
}
```

I think that the problem is in following condition where the `GetThread()` returns `nullptr`:   https://github.com/dotnet/runtime/blob/release/5.0/src/coreclr/src/vm/ceemain.cpp#L1864-L1865

**Comment by janvorli** [comment]
@ww898 thank you for all the details. The issue was caused by a bit different change. We were relying on DllMain being called on thread attach / detach before and then switched to the `thread_local`. The `__thread` doesn't support any constructors / destructors on objects at all.
So it seems the SIGSEGV is somehow related to an interaction between your profiler and our PAL. I wonder if you also have a thread exit handler registered that would end up being called before our one and do something that our one doesn't like.

**Comment by ww898** [comment]
I reproducted the issue (with some minor fixes in `build.sh` and `repro.sh`) on Linux x64 on .NET v6.0.100 only (.NET 5.0 work properly):
```
$ ./repro.sh 
Warning NETSDK1174: The abbreviation of -p for --project is deprecated. Please use --project.
Building...
Initialize profiler!
Launching ForceGC thread
Thread created, id=94dbea70
Thread created, id=94e13470
Thread created, id=94e325f0
Trying to force gc
Thread created, id=24000c60
GC forced
ForceGC threadId=24000c60
Thread created, id=94f56e30
Thread created, id=94f27cc0
Thread created, id=9501bfa0
Thread created, id=95021020
Thread created, id=140022b0
Thread created, id=94fdf9a0
Thread created, id=9517e790
Thread created, id=c001860
Thread created, id=9517d4d0
Thread created, id=16a0
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:5001
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /home/***/Downloads/Apps/macos_thread_profiler_repro-main/webapp
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread destroyed, id=94e325f0
Thread created, id=10000cd0
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread destroyed, id=10000cd0
Thread created, id=10000cd0
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread destroyed, id=10000cd0
Thread destroyed, id=9517d4d0
Thread destroyed, id=140022b0
Thread destroyed, id=16a0
Thread destroyed, id=9501bfa0
Thread destroyed, id=c001860
Thread created, id=aa90
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
^CThread created, id=20000c60
Thread destroyed, id=aa90
Calling EnumThreads because we have observed the GC thread being destroyed
Now enumerating threads
Saw ForceGC thread id in enum list!
Thread created, id=200025d0
^CThread created, id=20003bc0
Thread created, id=a2c0
Thread created, id=b740
Thread destroyed, id=20003bc0
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
Thread created, id=fc001e10
Thread created, id=ec0
Thread created, id=c003f50
Thread created, id=280590f0
Thread created, id=fc079240
Thread created, id=f00018d0
Thread created, id=fc07b000
Thread created, id=f40016a0
Thread destroyed, id=94f27cc0
```

**Comment by tommcdon** [comment]
@VSadov @davmason is this issue fixed?

**Comment by davmason** [comment]
It was fixed in #65638
